server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java

```diff
@@ -0,0 +1,291 @@
+          measure.setValue(durations.decode(valueAsString).toMinutes());
+          break;
+        case FLOAT:
+        case PERCENT:
+        case RATING:
+          typeValidations.validate(valueAsString, PropertyType.FLOAT.name(), null);
+          measure.setValue(Double.parseDouble(valueAsString));
+          break;
+        case LEVEL:
+          typeValidations.validate(valueAsString, PropertyType.METRIC_LEVEL.name(), null);
+          measure.setTextValue(valueAsString);
+        case STRING:
+        case DATA:
+        case DISTRIB:
+          measure.setTextValue(valueAsString);
+          break;
+        default:
+          throw new IllegalArgumentException("Unsupported metric type:" + metricType.description());
+      }
+    } catch (Exception e) {
+      throw new IllegalArgumentException(String.format("Ill formatted value '%s' for metric type '%s'", valueAsString, metricType.description()), e);
+    }
+  }
+
+  private MetricDto searchMetric(DbSession dbSession, Request request) {
+    Integer metricId = request.paramAsInt(PARAM_METRIC_ID);
+    String metricKey = request.param(PARAM_METRIC_KEY);
+    checkArgument(metricId != null ^ metricKey != null, "The metric id or the metric key must be provided, not both.");
+
+    if (metricId != null) {
+      return dbClient.metricDao().selectById(dbSession, metricId);
+    }
+
+    return dbClient.metricDao().selectByKey(dbSession, metricKey);
+  }
+
+  private ComponentDto searchProject(DbSession dbSession, Request request) {
+    String projectUuid = request.param(PARAM_PROJECT_ID);
+    String projectKey = request.param(PARAM_PROJECT_KEY);
+    checkArgument(projectUuid != null ^ projectKey != null, "The project key or the project id must be provided, not both.");
+
+    if (projectUuid != null) {
+      ComponentDto project = dbClient.componentDao().selectNullableByUuid(dbSession, projectUuid);
+      if (project == null) {
+        throw new NotFoundException(String.format("Project id '%s' not found", projectUuid));
+      }
+
+      return project;
+    }
+
+    ComponentDto project = dbClient.componentDao().selectNullableByKey(dbSession, projectKey);
+    if (project == null) {
+      throw new NotFoundException(String.format("Project key '%s' not found", projectKey));
+    }
+
+    return project;
+  }
+
+  private String measureValueDescription() {
+    StringBuilder description = new StringBuilder("Measure value. Value type depends on metric type:");
+    description.append("<ul>");
+    for (Metric.ValueType metricType : Metric.ValueType.values()) {
+      description.append("<li>");
+      description.append(String.format("%s - %s", metricType.description(), metricTypeWsDescription(metricType)));
+      description.append("</li>");
+    }
+    description.append("</ul>");
+
+    return description.toString();
+  }
+
+  private String metricTypeWsDescription(Metric.ValueType metricType) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "metricTypeWsDescription" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)